core/src/main/scala/kafka/log/AbstractIndex.scala

```diff
@@ -34,16 +34,37 @@ import scala.math.ceil
 /**
  * The abstract index class which holds entry format agnostic methods.
  *
- * @param file The index file
+ * @param _file The index file
  * @param baseOffset the base offset of the segment that this index is corresponding to.
  * @param maxIndexSize The maximum index size in bytes.
  */
-abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Long,
+abstract class AbstractIndex[K, V](@volatile var _file: File, val baseOffset: Long,
                                    val maxIndexSize: Int = -1, val writable: Boolean) extends Closeable with Logging {
 
+  def file: File = {
+    if (!initialized) {
+      inLock(lock) {
+        if (!initialized) {
+          initializeMmapAndLength
+        }
+      }
+    }
+    _file
+  }
+
+  def fileName: String = {
+    _file.getName
+  }
+
+  def file_=(f: File) {
```

Comment:
Hmm, this looks like something we already do but I wonder if it safe to switch to using a new `File` at any point in time? What happens to other state like `_length` and `_mmap` after the switch?

Reply:
That is a good point. I dug into the codes and found that we use the setter of `file` only when we rename the file (e.g. alter log dir, add suffix). In this case, only the metadata (i.e. inode) changes and the physical location of the file stays the same so I think the state like `_length` and `_map` are the same and we don't need to reinitialize them after the switch.